huaweicloud/services/er/data_source_huaweicloud_er_instances.go

```diff
@@ -0,0 +1,254 @@
+	if err != nil {
+		return nil, fmt.Errorf("error filting instance list: %s", err)
+	}
+
+	result := make([]instances.Instance, 0, len(filterResult))
+	tagFilter := d.Get("tags").(map[string]interface{})
+	for _, val := range filterResult {
+		item := val.(instances.Instance)
+		tagmap := utils.TagsToMap(item.Tags)
+
+		// Filter instances list by tags, if the filter is nil, skip and return all fileterResult elements.
+		if !utils.HasMapContains(tagmap, tagFilter) {
+			continue
+		}
+		result = append(result, item)
+	}
+	return result, nil
+}
+
+func flattenInstances(all []instances.Instance) []map[string]interface{} {
+	if len(all) < 1 {
+		return nil
+	}
+
+	result := make([]map[string]interface{}, len(all))
+	for i, instance := range all {
+		result[i] = map[string]interface{}{
+			"id":                                 instance.ID,
+			"asn":                                instance.ASN,
+			"name":                               instance.Name,
+			"description":                        instance.Description,
+			"status":                             instance.Status,
+			"enterprise_project_id":              instance.EnterpriseProjectId,
+			"tags":                               utils.TagsToMap(instance.Tags),
+			"created_at":                         instance.CreatedAt,
+			"updated_at":                         instance.UpdatedAt,
+			"enable_default_propagation":         instance.EnableDefaultPropagation,
+			"enable_default_association":         instance.EnableDefaultAssociation,
+			"auto_accept_shared_attachments":     instance.AutoAcceptSharedAttachments,
+			"default_propagation_route_table_id": instance.DefaultPropagationRouteTableId,
+			"default_association_route_table_id": instance.DefaultAssociationRouteTableId,
+			"availability_zones":                 instance.AvailabilityZoneIds,
+		}
+	}
+	return result
+}
+
+func buildSliceIgnoreEmptyElement(e string) []string {
+	if e != "" {
+		return []string{e}
+	}
+	return nil
+}
+
+func buildInstanceListOpts(d *schema.ResourceData) instances.ListOpts {
+	return instances.ListOpts{
+		EnterpriseProjectIds: buildSliceIgnoreEmptyElement(d.Get("enterprise_project_id").(string)),
+		Statuses:             buildSliceIgnoreEmptyElement(d.Get("status").(string)),
+		IDs:                  buildSliceIgnoreEmptyElement(d.Get("instance_id").(string)),
+		OwnedBySelf:          d.Get("owned_by_self").(bool),
+		SortKey:              []string{"name"},
+	}
+}
+
+func dataSourceInstancesRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+	client, err := cfg.ErV3Client(region)
+	if err != nil {
+		return diag.Errorf("error creating ER v3 client: %s", err)
+	}
+
+	resp, err := instances.List(client, buildInstanceListOpts(d))
+	if err != nil {
+		return diag.Errorf("error retrieving instances: %s", err)
+	}
+	if _, ok := d.GetOk("tags"); ok && d.Get("name") != "" {
```

Comment:
this judgement can be deleted as it's also checked in the function `filterInstances`